### PR TITLE
feat: Support moving a user and it's teams to an org as temporary approach

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?

Corresponding website PR: 
https://github.com/calcom/website/pull/638

## Type of change
- [ ] New feature (non-breaking change which adds functionality)

## How should this be tested?
- Have a user within a team with many apps installed and many event-types across teams
- Migrate the user to an org
- Check that all the teams and event-types and apps are accessible.

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist
- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/pages/[user].tsx [hunk 1]
```diff
@@ -312,6 +316,18 @@ export const getServerSideProps: GetServerSideProps<UserPageProps> = async (cont
     avatar: `/${user.username}/avatar.png`,
   }));
 
+  if (!isOrgContext) {
+    const redirect = await getTemporaryOrgRedirect({
+      slug: usernameList[0],
```
Comment (hariombalhara) [inline apps/web/pages/[user].tsx:321]: Not supporting redirect for dynamic booking link for now as they would most probably be not hardcoded anywhere.

File: apps/web/pages/[user]/[type].tsx [hunk 2]
```diff
@@ -74,6 +77,20 @@ async function getDynamicGroupPageProps(context: GetServerSidePropsContext) {
     context.params?.orgSlug
   );
 
+  const isOrgContext = currentOrgDomain && isValidOrgDomain;
+
+  if (!isOrgContext) {
+    const redirect = await getTemporaryOrgRedirect({
+      slug: usernames[0],
+      redirectType: RedirectType.User,
+      eventTypeSlug: slug,
```
Comment (hariombalhara) [inline apps/web/pages/[user]/[type].tsx:86]: Redirect for user even-type page.

File: apps/web/pages/team/[slug].tsx [hunk 3]
```diff
@@ -272,13 +275,28 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
...
     slug: slugify(slug ?? ""),
     orgSlug: currentOrgDomain,
     isTeamView: true,
     isOrgView: isValidOrgDomain && context.resolvedUrl === "/",
   });
+
+  if (!isOrgContext && slug) {
+    const redirect = await getTemporaryOrgRedirect({
+      slug: slug,
+      redirectType: RedirectType.Team,
```
Comment (hariombalhara) [inline apps/web/pages/team/[slug].tsx:291]: Redirect for team profile page.

File: apps/web/pages/team/[slug]/[type].tsx [hunk 4]
```diff
@@ -75,6 +78,19 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
     context.req.headers.host ?? "",
     context.params?.orgSlug
   );
+  const isOrgContext = currentOrgDomain && isValidOrgDomain;
+
+  if (!isOrgContext) {
+    const redirect = await getTemporaryOrgRedirect({
+      slug: teamSlug,
+      redirectType: RedirectType.Team,
+      eventTypeSlug: meetingSlug,
```
Comment (hariombalhara) [inline apps/web/pages/team/[slug]/[type].tsx:87]: Redirect for team event.

File: packages/trpc/server/routers/viewer/organizations/list.handler.ts [hunk 5]
```diff
@@ -29,6 +29,13 @@ export const listHandler = async ({ ctx }: ListHandlerInput) => {
     },
   });
 
+  if (!membership) {
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/organizations/list.handler.ts:32]: Without membership wouldn't be able to see listing of an orgzniation

File: apps/web/lib/getTemporaryOrgRedirect.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  const prisma = (await import("@calcom/prisma")).default;
+  log.debug(
+    `Looking for redirect for`,
+    safeStringify({
+      slug,
+      redirectType,
+      eventTypeSlug,
+    })
+  );
+  // Not supporting dynamic group for now
```
Comment (hariombalhara) [inline apps/web/lib/getTemporaryOrgRedirect.ts:24]:
```suggestion
```

File: apps/web/lib/getTemporaryOrgRedirect.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+  const prisma = (await import("@calcom/prisma")).default;
+  log.debug(
+    `Looking for redirect for`,
+    safeStringify({
+      slug,
+      redirectType,
+      eventTypeSlug,
+    })
+  );
+  const redirect = await prisma.tempOrgRedirect.findUnique({
```
Comment (hariombalhara) [inline apps/web/lib/getTemporaryOrgRedirect.ts:24]: It's a temporary table and would be really lightweight. So, It should be quick to query but there would still be DB latency that would be added in querying a booking page.

I feel it's okay to have it for now, till we actually see a significant issue with it(as we will remove it anyway soon when mono-user Org schema is ready)

If we see an issue, our option then is to query this in parallel, with other queries on Booking Page which is possible

File: apps/web/pages/org/[orgSlug]/[user]/[type]/embed.tsx [hunk 8]
```diff
@@ -1,19 +1,7 @@
-import type { GetServerSidePropsContext } from "next";
+import withEmbedSsr from "@lib/withEmbedSsr";
 
 import { getServerSideProps as _getServerSideProps } from "../[type]";
 
 export { default } from "../[type]";
 
-export const getServerSideProps = async (context: GetServerSidePropsContext) => {
```
Comment (hariombalhara) [inline apps/web/pages/org/[orgSlug]/[user]/[type]/embed.tsx:7]: I am not sure why it wasn't using withEmbedSsr, all other embed pages were using that only.

**Review by zomars — COMMENTED** [review]
Left some comments in the website PR. Also. Shouldn't this live in the web app admin panel instead of the website? (even if it's temporary)

**Comment by hariombalhara** [comment]
Thanks for the review @zomars 🙏

There are 2 reasons why I didn't put into Admin Panel
- It doesn't concern our self-hosters at all because it is something only we are going to need and that too temporarily.
- More importantly, there is an endpoint in there that I don't want to be exposed to people. That endpoint that allow people to join an org by themselves(Though that logic is explicitly limited to isAdmin at the moment) and that could be abused.

**Review by zomars — APPROVED** [review]
Concerns were addressed 🙏🏽

**Pull request merged**